### PR TITLE
[Placement group] Scheduler map refactoring part 1.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

To support atomic creation of placement group, which will add complexity to codebase, we'd like to have clean scheduler code. This PR simplifies management of `node_to_leased_bundles` and ` placement_group_to_bundles` in-memory map, which are reverse index to each other by creating a class that updates both map at the same time with clean interface.

The next PR will refactor all leasing related field into `LeasingContext` class.

## Related issue number

https://github.com/ray-project/ray/issues/9807

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 1]
```diff
@@ -130,6 +131,71 @@ class GcsStrictSpreadStrategy : public GcsScheduleStrategy {
...
+  ///
+  /// \param placement_group_id Placement group id
+  /// \return True if succeed. False otherwise.
+  bool Erase(const PlacementGroupID &placement_group_id);
+
+  /// Get BundleLocation of placaement group id.
+  ///
+  /// \param placement_group_id Placement group id of this bundle locations.
+  /// \return Bundle locations that are associated with a given placement group id.
+  const absl::optional<std::shared_ptr<BundleLocations>> Get(
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:170]:
```suggestion
  const absl::optional<std::shared_ptr<BundleLocations>> GetBundleLocations(
```

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 2]
```diff
@@ -130,6 +131,71 @@ class GcsStrictSpreadStrategy : public GcsScheduleStrategy {
...
+  /// \return True if succeed. False otherwise.
+  bool Erase(const ClientID &node_id);
+
+  /// Erase bundle locations associated with a given placement id.
+  ///
+  /// \param placement_group_id Placement group id
+  /// \return True if succeed. False otherwise.
+  bool Erase(const PlacementGroupID &placement_group_id);
+
+  /// Get BundleLocation of placaement group id.
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:166]:
```suggestion
  /// Get BundleLocation of placement group id.
```

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 3]
```diff
@@ -130,6 +131,71 @@ class GcsStrictSpreadStrategy : public GcsScheduleStrategy {
...
+  /// \param placement_group_id Placement group id of this bundle locations.
+  /// \return Bundle locations that are associated with a given placement group id.
+  const absl::optional<std::shared_ptr<BundleLocations>> Get(
+      const PlacementGroupID &placement_group_id);
+
+  /// Get BundleLocation of node id.
+  ///
+  /// \param placement_group_id Node id of this bundle locations.
+  /// \return Bundle locations that are associated with a given node id.
+  const absl::optional<std::shared_ptr<BundleLocations>> Get(const ClientID &node_id);
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:177]:
```suggestion
  const absl::optional<std::shared_ptr<BundleLocations>> GetBundlesOnNode(const ClientID &node_id);
```

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 4]
```diff
@@ -130,6 +131,71 @@ class GcsStrictSpreadStrategy : public GcsScheduleStrategy {
...
+
+  /// Update the index to contain new node information. Should be used only when new node
+  /// is added to the cluster.
+  ///
+  /// \param alive_nodes map of alive nodes.
+  void UpdateNewNodes(
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:183]:
```suggestion
  void AddNodes(
```

Comment (ericl) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:183]: Or, can we add them one at a time?

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 5]
```diff
@@ -130,6 +131,71 @@ class GcsStrictSpreadStrategy : public GcsScheduleStrategy {
...
 };
 
+/// A data structure that encapsulates information regarding bundle resource leasing
+/// status.
+class LeasingContext {
+  // TODO(sang): Implement in the next PR.
+};
+
+/// A data structure that helps fast bundle location lookup.
+class BundleLocationIndex {
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:141]: Consider unit testing this data structure.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc [hunk 6]
```diff
@@ -411,41 +398,126 @@ void GcsPlacementGroupScheduler::OnAllBundleSchedulingRequestReturned(
...
+  return true;
+}
+
+const absl::optional<std::shared_ptr<BundleLocations>> BundleLocationIndex::Get(
+    const PlacementGroupID &placement_group_id) {
+  auto it = placement_group_to_bundle_locations_.find(placement_group_id);
+  if (it == placement_group_to_bundle_locations_.end()) {
+    return {};
+  }
+  return it->second;
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:501]: Can we return a const pointer to indicate the caller should not mutate it?

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc [hunk 7]
```diff
@@ -411,41 +398,126 @@ void GcsPlacementGroupScheduler::OnAllBundleSchedulingRequestReturned(
...
+}
+
+const absl::optional<std::shared_ptr<BundleLocations>> BundleLocationIndex::Get(
+    const ClientID &node_id) {
+  auto it = node_to_leased_bundles_.find(node_id);
+  if (it == node_to_leased_bundles_.end()) {
+    return {};
+  }
+  return it->second;
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:510]: Same question on const.

**Review by ericl — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 8]
```diff
@@ -130,6 +131,71 @@ class GcsStrictSpreadStrategy : public GcsScheduleStrategy {
...
+  /// Get BundleLocation of placaement group id.
+  ///
+  /// \param placement_group_id Placement group id of this bundle locations.
+  /// \return Bundle locations that are associated with a given placement group id.
+  const absl::optional<std::shared_ptr<BundleLocations>> Get(
+      const PlacementGroupID &placement_group_id);
+
+  /// Get BundleLocation of node id.
+  ///
+  /// \param placement_group_id Node id of this bundle locations.
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:175]:
```suggestion
  /// \param node_id Node id of this bundle locations.
```

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 9]
```diff
@@ -227,32 +293,22 @@ class GcsPlacementGroupScheduler : public GcsPlacementGroupSchedulerInterface {
...
   /// It is required to know if placement group has been removed or not.
   absl::flat_hash_set<PlacementGroupID> placement_group_leasing_in_progress_;
 
-  /// A map from placement group id to bundle locations.
-  /// It is used to destroy bundles for the placement group. When we reschedule bundles,
-  /// we can get the location of other bundles from here.
-  /// NOTE: It is a reverse index of `node_to_leased_bundles`.
-  absl::flat_hash_map<PlacementGroupID, std::shared_ptr<BundleLocations>>
-      placement_group_to_bundle_locations_;
+  /// Index to lookup bundle locations of node or placement group
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:310]:
```suggestion
  /// Index to lookup bundle locations of node or placement group.
```

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 10]
```diff
@@ -130,6 +131,71 @@ class GcsStrictSpreadStrategy : public GcsScheduleStrategy {
...
+  void AddBundleLocations(const PlacementGroupID &placement_group_id,
+                          std::shared_ptr<BundleLocations> bundle_locations);
+
+  /// Erase bundle locations associated with a given node id.
+  ///
+  /// \param node_id Placement group id
+  /// \return True if succeed. False otherwise.
+  bool Erase(const ClientID &node_id);
+
+  /// Erase bundle locations associated with a given placement id.
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:160]:
```suggestion
  /// Erase bundle locations associated with a given placement group id.
```

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 11]
```diff
@@ -130,6 +131,71 @@ class GcsStrictSpreadStrategy : public GcsScheduleStrategy {
...
+  ///
+  /// \param placement_group_id
+  /// \param bundle_locations Bundle locations that will be associated with the placement
+  /// group id.
+  void AddBundleLocations(const PlacementGroupID &placement_group_id,
+                          std::shared_ptr<BundleLocations> bundle_locations);
+
+  /// Erase bundle locations associated with a given node id.
+  ///
+  /// \param node_id Placement group id
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:156]:
```suggestion
  /// \param node_id The id of node.
```

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc [hunk 12]
```diff
@@ -411,41 +398,126 @@ void GcsPlacementGroupScheduler::OnAllBundleSchedulingRequestReturned(
...
+
+bool BundleLocationIndex::Erase(const ClientID &node_id) {
+  const auto leased_bundles_it = node_to_leased_bundles_.find(node_id);
+  if (leased_bundles_it == node_to_leased_bundles_.end()) {
+    return false;
+  }
+
+  const auto &bundle_locations = leased_bundles_it->second;
+  for (const auto &bundle_location : *bundle_locations) {
+    // Remove corresponding placement group id.
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:455]: Sorry, I didn't find the code.

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:455]: No worries haha. I was surprised I couldn't find this problem from the PR before!

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h (hunk 9 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:310]: Thanks for all these catch!!! Will be more careful about commenting next time.

**Review by ffbin — APPROVED** [review]
LGTM

**Pull request merged**